tests/auth/ldap_test.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+"""Test ldap
+
+:copyright: Copyright (c) 2023 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+import pytest
+from pykern.pkdebug import pkdlog, pkdp
+
+
+def test_happy_path():
+    from pykern.pkcollections import PKDict
+    from pykern import pkunit
+
+    res = PKDict(
+        email="vagrant@radiasoft.net",
+        password="vagrant",
+    )
+    r, u = _call_login(res)
+    pkunit.pkeq(200, r.status_as_int())
+    pkunit.pkre("location.*/complete-registration", r.content_as_str())
+    pkunit.pkne(None, u)
+
+
+def test_cred_validation():
+    from pykern.pkcollections import PKDict
+    from pykern import pkunit
+
+    res = PKDict(
+        email="any-user",
+        password="",
+    )
+    r = _call_login(res)[0]
+    pkunit.pkok("form_error" in r.content_as_str(), "did not return form_error")
```

Comment:
This is pointing out an interesting problem for our tests when calling apis directly. What's missing is comparing json to a struct. In this case, we want something like this:

```py
pkunit.pkre("invalid user and", r.data.form_error)
```

You don't have this, but how would you get it?

BTW, you generally do not want a message, because this obscures the values. The nice thing pkunit is that it outputs the line and file where the error occurs so you can just jump to that part of the test to see what is wrong. In general, pkok should be avoided for pkre, because the latter shows what the re and the value in the error message, which is useful for debugging the test. The actual error location is output so strings like "did not return form_error" aren't all that useful.

Reply:
> What's missing is comparing json to a struct. In this case, we want something like this: `pkunit.pkre("invalid user and", r.data.form_error)`

What's an example of a JSON and struct we would want to compare for testing?

I'm aware of the specific struct type in C, what is a struct in this context? Do we refer to PKDict as a struct, and therefore the _SReply we get from calling api's as well?

My basic understanding is we use JSON for application/code schema and storing/sharing simulation data (I think). In this case the expected is just a string "invalid user...", what JSON would need to be referenced in authentication testing (or other testing)?

May not be good questions, feels like I'm missing context so trying to orient my understanding of what's being discussed.

Reply:
There's actually no good word for the result of a JSON deserialization. JSON is a string. We could just call it an "object", which it is, but that's not too descriptive. Calling it a PKDict is also incorrect as it could be a list or a string, etc. That's why I used "struct". Sorry for the confusion.